inference.py

```diff
@@ -1,128 +1,97 @@
-    elif message["role"] == "assistant":
-        text = "assistant:\n{content}\n".format(content=message.get("content", ""))
-
-    else:
-        raise ValueError(f'Unsupported role: {message["role"]}')
-
-    input_ids = tokenizer(text, add_special_tokens=False, return_tensors="pt").input_ids.to("cuda:0")
-    return input_ids
+def to_tokens(message: TurnMessage, tokenizer: LlamaTokenizer):
+    text = str(message)
+    return tokenizer(text, add_special_tokens=False, return_tensors="pt").input_ids.to(
+        "cuda:0"
+    )
 
 
-def prepare_messages_for_inference(tokenizer, messages, functions=None):
+def prepare_messages_for_inference(
+    tokenizer: LlamaTokenizer, messages: List[TurnMessage], functions=None
+):
     all_messages = []
     if functions is not None:
-        all_messages.append({"role": "system", "content": generate_schema_from_functions(functions)})
-    all_messages.append({"role": "system", "content": SYSTEM_MESSAGE})
+        all_messages.append(
+            TurnMessage(
+                role="system", content=generate_schema_from_functions(functions)
+            )
+        )
+    all_messages.append(TurnMessage(role="system", content=SYSTEM_MESSAGE))
+
     for message in messages:
-        if message.get("role") == "assistant":
-            if message.get("content"):
-                all_messages.append({"role": "assistant", "content": message.get("content")})
-            if message.get("function_call"):
-                all_messages.append({"role": "assistant", 
-                                     "to": "functions." + message.get("function_call", {}).get("name"), 
-                                     "content": message.get("function_call", {}).get("arguments")})
-        elif message.get("role") == "function":
-            message["name"] = "functions." + message.get("name", "")
-            all_messages.append(message)
-        else:
-            all_messages.append(message)
-            
-    all_messages.append({"role": "assistant", "content": None})
-    #print(all_messages)
-    all_input_ids = [prepare_message_for_inference(tokenizer, msg) for msg in all_messages]
+        if message.role == "assistant":
+            if message:
+                all_messages.append(
+                    TurnMessage(role="assistant", content=message.content)
+                )
+            if message.function_call:
+                fc = message.function_call
+                all_messages.append(
+                    TurnMessage(
+                        role="assistant",
+                        to=f"functions.{fc.name}",
+                        content=fc.arguments,
+                    )
+                )
+        elif message.role == "function":
+            all_messages.append(
+                TurnMessage(
+                    role="function",
+                    name=f"functions.{message.name}",
+                    content=message.content,
+                )
+            )
+        all_messages.append(message)
```

Comment:
This should be in else block, as this will duplicate function and assistant messages